bazel_tools/proto.bzl

```diff
@@ -159,3 +165,95 @@ proto_gen = rule(
+        return []
+
+def proto_jars(
+        name,
+        srcs,
+        strip_import_prefix = "",
+        deps = [],
+        proto_deps = [],
+        java_deps = [],
+        scala_deps = [],
+        file_root = None,
+        javadoc_root_packages = [],
+        maven_group = None,
+        maven_artifact_prefix = None,
+        maven_java_artifact_suffix = "java-proto"):
+    # Tarball containing the *.proto files.
+    pkg_tar(
+        name = "%s_src" % name,
+        srcs = srcs,
+        extension = "tar.gz",
+        strip_prefix = strip_import_prefix,
+        package_dir = file_root,
+        visibility = ["//visibility:public"],
+    )
+
+    # Compiled protobufs. Used in subsequent targets.
+    proto_library(
+        name = name,
+        srcs = srcs,
+        strip_import_prefix = strip_import_prefix,
+        visibility = ["//visibility:public"],
+        deps = deps + proto_deps,
+    )
+
+    # JAR containing the generated Java bindings.
+    native.java_proto_library(
+        name = "%s_java" % name,
+        tags = maven_tags(maven_group, maven_artifact_prefix, maven_java_artifact_suffix),
+        visibility = ["//visibility:public"],
+        deps = [":%s" % name] + java_deps,
+    )
+
+    if maven_group and maven_artifact_prefix:
+        pom_file(
+            name = "%s_java_pom" % name,
+            target = ":%s_java" % name,
+            visibility = ["//visibility:public"],
+        )
+
+    if javadoc_root_packages:
+        javadoc_library(
+            name = "%s_java_javadoc" % name,
+            srcs = [":%s_java" % name],
+            root_packages = javadoc_root_packages,
+            visibility = ["//visibility:public"],
+            deps = ["@maven//:com_google_protobuf_protobuf_java"],
+        ) if not is_windows else None
+    else:
+        # Create an empty Javadoc JAR for uploading proto JARs to Maven Central.
+        # We don't need to create an empty JAR file for sources, because `java_proto_library`
+        # creates a source JAR automatically.
+        pkg_empty_zip(
+            name = "%s_java_javadoc" % name,
+            out = "%s_java_javadoc.jar" % name,
+        )
+
+    # JAR containing the generated Scala bindings.
+    proto_gen(
+        name = "%s_scala_sources" % name,
+        srcs = [":%s" % name],
+        plugin_exec = "//scala-protoc-plugins/scalapb:protoc-gen-scalapb",
+        plugin_name = "scalapb",
+        visibility = ["//visibility:public"],
+        deps = deps + proto_deps,
+    )
+
+    scala_library(
+        name = "%s_scala" % name,
+        srcs = [":%s_scala_sources" % name],
+        unused_dependency_checker_mode = "error",
+        visibility = ["//visibility:public"],
+        deps = [
+            "@maven//:com_google_protobuf_protobuf_java",
+            "@maven//:com_thesamet_scalapb_lenses_2_12",
+            "@maven//:com_thesamet_scalapb_scalapb_runtime_2_12",
+        ] + ["%s_scala" % label for label in proto_deps] + scala_deps,
+    )
```

Comment:
Did you intend to switch `//ledger-api/grpc-definitions` over to this? I can’t find any place where you use this atm.

Reply:
I do, but I actually did this so I could use Scala bindings for the KVUtils proto.

Reply:
(In another, future piece of work.)

Reply:
I would have switched it over in this PR so this isn’t just unused code but I’ll leave it up to you if you want to include it here or do it in a separate PR.

Reply:
It's pretty much done but I wanted to keep this a little smaller. (That's a big one in its own right. Expect it this morning.)